### PR TITLE
Add action to decommission legacy monitoring cluster alerts

### PR DESCRIPTION
**Pull request by jbaiera**
This follows on from #62668 by adding an action that will proactively remove any watches that monitoring has configured. The action toggles on the new setting that informs the cluster to tear down any previously created cluster alerts, and after that is accepted, the action immediately attempts a best-effort refresh of cluster alert resources in order to force their removal in case collection is disabled or delayed.

Since resources are controlled lazily by the existing monitoring exporters, extra care was taken to ensure that any in-flight resource management operations do not race against any resource actions taken by the migration action. Resource installation code was updated with callbacks to report any errors instead of just logging them.

### CODE REVIEW

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+     * Executed after the settings update has been accepted, this collects the enabled and disabled exporters, requesting each of them
+     * to explicitly remove their installed alerts if possible. This makes sure that alerts are removed in a timely fashion instead of
+     * waiting for metrics to be bulked into the monitoring cluster.
+     */
+    private ActionListener<ClusterUpdateSettingsResponse> afterSettingUpdate(ActionListener<MonitoringMigrateAlertsResponse> listener) {
+        return ActionListener.wrap(clusterUpdateSettingsResponse -> {
+
+            // Ensure positive result
+            if (!clusterUpdateSettingsResponse.isAcknowledged()) {
+                listener.onFailure(new ElasticsearchException("Failed to update monitoring migration settings"));
```
Comment (jakelandis) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java:86]: nit: include the setting name in the error.
nit: prefer `== false` over `!`

Comment (jakelandis) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java:86]: If the setting is already true, does clusterUpdateSettingResponse.isAcknoledged() still return true ?

**Review by jakelandis — COMMENTED** [review]
Getting a bit lost in the introduction of more state via `ExporterResourceStatus` . 

It seems to add quite a bit of complexity having multiple possible deploy states. Under what circumstances wouldn't your code to do the main work on the success of setting MIGRATION_DECOMMISSION_ALERTS be sufficient to avoid race conditions ?  Is there a reasonable trade off here, such that we could simplify the code to only check a single state and error early via the migrate REST endpoint if not in that state (or obtain that state in a reasonable amount of time) ?

EDIT: for example, would it be possible to try to aquire (with a timeout) the semaphore from `MonitoringExecution` effectively pausing monitoring while we do the migration avoiding the race conditions. Since this is a one time call...I think that would be an OK tradeoff. Monitoring is generally no faster then every 10s, and in a healthy system executes pretty quickly. 

EDIT2: ^^ that won't actually work exactly as describe due to the "router" also using the exporters...but the idea is the same...is there a trade off where we can make the use of the contended resources mutually exclusive to avoid complexity of managing so much state ?

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java (hunk 1 above)
Comment (jbaiera) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java:102]: I added a test that runs the migration multiple times. No failures encountered.

**Comment by jbaiera** [comment]
I've gone through and made some changes. There is now a migration coordinator object that is acquired when a migration action begins. Every exporter checks this coordinator before running its installation steps. If the coordinator is acquired already, the exporter blocks as if its resources are not ready yet. Once the migration operations complete, the migration action releases the coordinator, and exporters are able to perform their installation tasks as normal.

Since we no longer have to worry about exporters running their installation tasks DURING a migration, I've removed the retry code and `IN_PROGRESS` states. If an exporter is ever in progress during a migration (should be impossible with the coordinator in place), I have it fail the migration for sanity's sake.

**Comment by jbaiera** [comment]
@elasticmachine update branch

**Review by jakelandis — COMMENTED** [review]
a few comments

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        // First, enable the migration coordinator block
+        if (migrationCoordinator.tryBlockInstallationTasks(ALERT_MIGRATION_INSTALLATION_WAIT_TIMEOUT) == false) {
+            throw new EsRejectedExecutionException("Could not migrate cluster alerts. Timed out waiting to block resource operations.");
+        }
+        // Wrap the listener to unblock resource installation before completing
+        listener = ActionListener.runBefore(listener, migrationCoordinator::unblockInstallationTasks);
+        try {
+            Settings.Builder decommissionAlertSetting = Settings.builder().put(Monitoring.MIGRATION_DECOMMISSION_ALERTS.getKey(), true);
+            client.admin().cluster().prepareUpdateSettings().setPersistentSettings(decommissionAlertSetting)
+                .execute(afterSettingUpdate(listener));
```
Comment (jakelandis) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java:85]: What thread is all this work happening on ?

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        this.client = client;
+        this.migrationCoordinator = migrationCoordinator;
+        this.exporters = exporters;
+    }
+
+    @Override
+    protected void masterOperation(Task task, MonitoringMigrateAlertsRequest request, ClusterState state,
+                                   ActionListener<MonitoringMigrateAlertsResponse> listener) throws Exception {
+        // First, enable the migration coordinator block
+        if (migrationCoordinator.tryBlockInstallationTasks(ALERT_MIGRATION_INSTALLATION_WAIT_TIMEOUT) == false) {
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java:77]: Actually... after looking at this abit more, I think the semaphore should be replaced by a simple atomic boolean. Here we can throw an exception immediately if multiple requests come in while processing, and the other places are currently using this effectively as a boolean .

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java (hunk 2 above)
Comment (jbaiera) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java:85]: Hmm, if the master node it seems that this runs on a master service thread that completed the settings update. If not, I believe it's a network thread. I'll fork it over to management after the request succeeds.

**Comment by sgrodzicki** [comment]
@jbaiera what would be the ETA for this? We want to assess whether we will be able to make changes in Kibana for 7.11.

**Comment by jbaiera** [comment]
@sgrodzicki Sorry for the late response, been on vacation for a bit. This should be good to go once it passes review, which I hope shouldn't be much longer.

**Comment by jbaiera** [comment]
@elasticmachine update branch

**Comment by jbaiera** [comment]
@elasticmachine update branch

**Comment by jbaiera** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by jbaiera** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by jakelandis** [comment]
@elasticmachine update branch

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        this.migrationCoordinator = migrationCoordinator;
+        this.exporters = exporters;
+    }
+
+    @Override
+    protected void masterOperation(Task task, MonitoringMigrateAlertsRequest request, ClusterState state,
+                                   ActionListener<MonitoringMigrateAlertsResponse> listener) throws Exception {
+        // First, enable the migration coordinator block
+        if (migrationCoordinator.tryBlockInstallationTasks() == false) {
+            throw new EsRejectedExecutionException("Could not migrate cluster alerts. Timed out waiting to block resource operations.");
```
Comment (jakelandis) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java:73]: nit: there is no more time out.  

Also, can we change the language from "block" to "currently running"  ?

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/MonitoringMigrationCoordinator.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.monitoring.exporter;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A shared coordination object for blocking execution of exporters when a migration that involves them is in progress
```
Comment (jakelandis) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/MonitoringMigrationCoordinator.java:12]: nit: no longer blocking

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/Monitoring.java [hunk 6]
```diff
@@ -103,10 +106,12 @@ public Monitoring(Settings settings) {
         final ClusterSettings clusterSettings = clusterService.getClusterSettings();
         final CleanerService cleanerService = new CleanerService(settings, clusterSettings, threadPool, getLicenseState());
         final SSLService dynamicSSLService = getSslService().createDynamicSSLService();
+        final MonitoringMigrationCoordinator migrationCoordinator = new MonitoringMigrationCoordinator();
```
Comment (jakelandis) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/Monitoring.java:109]: This instance is passed to the exporters, but I don't think this is the instance that passed into the Transport action (that one is created by Guice and injected). Which would mean that any state set by the transport actions would not be read by the exporters, which defeats the purpose of having an external coordinator.  

I wonder if we should just remove the MonitoringMigrationCoordinator class entirely in favor of boolean in the tranport action that dis-allows running that action overlapping. 

If I understand the global coordinator is there to ensure that our cluster state updates from these operations don't end up re-constructing the exporters we are trying to disable while we are disabling them ? If so, can the transport action write a small piece of cluster state and the exporters onClusterState change read that and skip reconstruction while that state exists in the cluster ?

Comment (jakelandis) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/Monitoring.java:109]: OK ... I missed that that Node.java does `  pluginComponents.stream().forEach(p -> b.bind((Class) p.getClass()).toInstance(p));` where plugin components are those items created by the plugins, thereby making them the same instance objects that get injected. Sorry for the noise.

**Review by jakelandis — APPROVED** [review]
LGTM  (a couple nitpicks around naming), thanks for the iterations.

**Pull request merged**